File: 152-MaximumProductSubarray/152-MaximumProductSubarray.cpp
// Last updated: 6/11/2026, 11:35:35 AM
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int ans = *(max_element(nums.begin(), nums.end()));
        int n = nums.size();

        int maxp = 1;
        int minp = 1;
        for(int i=0; i<n; i++){
            maxp = maxp*nums[i];
            minp = minp*nums[i];
            ans = max({ans, maxp, minp});
            if(minp>maxp){
                swap(maxp, minp);
            }
            if(maxp <= 0){
                maxp = 1;
            }
            if(minp>=0){
                minp = 1;
            }
        }

        return ans;

    }
};

File: 152-MaximumProductSubarray/152-MaximumProductSubarray.py
# Last updated: 6/11/2026, 11:35:41 AM
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        max_prod = 1
        min_prod = 1
        res = max(nums)
        for i in nums:
            if(i==0):
                max_prod = 1
                min_prod = 1
                continue

            temp = i*max_prod
            max_prod = max(i*max_prod,i*min_prod,i)
            min_prod = min(temp,i*min_prod,i)
            res = max(res,max_prod)


        return res

